modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -52,18 +55,40 @@ def wait_computations_if_benchmark_mode(func):
     """
     if BenchmarkMode.get():
 
+        def extract_waitable(obj):
+            """Return an iterable of waitable objects contained in the `obj`."""
+            if hasattr(obj, "wait"):
+                return (obj,)
+            # fastpath for NumPy array of partitions
+            elif (
+                isinstance(obj, np.ndarray)
+                and obj.dtype == np.object
+                and len(obj) > 0
+                and hasattr(obj[np.unravel_index(0, shape=obj.shape)], "wait")
+            ):
+                return obj.flatten()
+            elif is_list_like(obj):
+                return tuple(chain(*(extract_waitable(o) for o in obj)))
+            else:
+                return tuple()
+
         @wraps(func)
         def wait(*args, **kwargs):
             """Wait for computation results."""
             result = func(*args, **kwargs)
-            if isinstance(result, tuple):
-                partitions = result[0]
+            if result is None:
+                # the operation was applied on partitions inplace so extracting those from arguments
+                partitions = extract_waitable((args, kwargs.values()))
```

Comment:
Why is this flexibility needed? Can we require a partition manager that any of its functions (where applicable) always return an array of partitions?

Reply:
It's needed just to extend a set of use cases. As we discover errors (inplace method was wrapped by the decorator) only in runtime it's better to be oversecured.

Reply:
if now we do not use it, then in this case it is enough to throw an exception. Writing code for the future is not a good idea.

Reply:
Okay, removed support for inplace functions in the decorator

Reply:
@dchigarev thanks :)